pkg/artifacts/signable.go

```diff
@@ -257,6 +268,104 @@ func extractTargetFromResults(obj objects.TektonObject, identifierSuffix string,
 	return ss
 }
 
+// ExtractStructuredTargetFromResults extracts structured signable targets aim to generate intoto provenance as materials within TaskRun results and store them as StructuredSignable.
+// categoryMarker can either be ARTIFACT_INPUTS or ARTIFACT_OUTPUTS.
+func ExtractStructuredTargetFromResults(obj objects.TektonObject, categoryMarker string, logger *zap.SugaredLogger) []*StructuredSignable {
+	objs := []*StructuredSignable{}
+	for _, res := range obj.GetResults() {
+		if strings.HasSuffix(res.Name, categoryMarker) {
+			if res.Value.ObjectVal["uri"] == "" {
+				logger.Errorf("%s should have uri key", res.Name)
+				continue
+			}
+			if res.Value.ObjectVal["digest"] == "" {
+				logger.Errorf("%s should follow digest key", res.Name)
+				continue
+			}
+			if err := checkDigest(res.Value.ObjectVal["digest"]); err != nil {
+				logger.Errorf("error getting digest %s: %v", res.Value.ObjectVal["digest"], err)
+				continue
+			}
+			if res.Value.ObjectVal["uri"] == "" {
+				logger.Errorf("URI cannot be empty for %s", res.Name)
+				continue
+			}
+			objs = append(objs, &StructuredSignable{URI: res.Value.ObjectVal["uri"], Digest: res.Value.ObjectVal["digest"]})
+		}
+	}
+	return objs
+}
+
+// MergeStructuredesultsIntoMaterials merge structure.
+func MergeStructuredResultsIntoMaterials(mats *[]slsa.ProvenanceMaterial, obj objects.TektonObject, categoryMarker string, logger *zap.SugaredLogger) {
+	// Retrieve structured provenance for inputs.
+	ssts := extractStructuredTargetFromResults(obj, ArtifactsInputsResultName, logger)
+	for _, s := range ssts {
+		splits := strings.Split(s.Digest, ":")
+		alg := splits[0]
+		digest := splits[1]
+		*mats = append(*mats, slsa.ProvenanceMaterial{
```

Comment:
I'm usually not a big fan of doing pointer mutation. Can we just return the materials list and let the caller do the append themselves?

Reply:
SG! I just return a list and let caller to merge them 🍡